File: Problems/No.19_Remove_Nth_Node_From_End_of_List/remove_nth_node_from_end_of_list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        
        # [Optimal Solution]
        # use 2 pointers, first iterate n and this means the difference
        
        # [My solution]
        # store some parameters
        curr = head
        sz = 1
        
        # preprocess (calculate the total size)
        while (curr.next != None):
            curr = curr.next
            sz += 1
        
        # edge case
        if (sz == 1):
            return None
        elif (sz == n):
            return head.next
        
        # start compute
        curr = head
        for i in range(sz-n-1):
            curr = curr.next
        curr.next = curr.next.next
        
        return head

File: Problems/No.19_Remove_Nth_Node_From_End_of_List/remove_nth_node_from_end_of_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        // [Optimal Solution]
        // use 2 pointers, first iterate n and this means the difference
        
        // [My solution]
        // store some parameters
        int sz = 1;
        ListNode* curr = head;
        
        // preprocess (calculate the total size)
        while (curr->next != NULL) {
            
            curr = curr->next;
            sz++;
            
        }
        
        // edge case
        if (sz == 1) return NULL;
        else if (sz == n) return head->next;
        
        // start compute
        curr = head;
        for (int i=0; i<sz-n-1; i++) curr = curr->next;
        curr->next = curr->next->next;
        
        return head;
        
    }
};
